### PR TITLE
o/snapstate: fix misalignment in RevisionOptions when requesting refresh candidates

### PR DESCRIPTION
**Pull request by ZeyadYasser**
It seems that `revOpts` here [overlord/snapstate/storehelpers.go#L662](https://github.com/snapcore/snapd/blob/0a39bc4fb1569007283e1e093dd8f30b7e801629/overlord/snapstate/storehelpers.go#L662) is not being built as expected due the sorting done above.

It looks like this is the only place that actually uses `revOpts` [overlord/snapstate/snapstate.go#L1601](https://github.com/snapcore/snapd/blob/0a39bc4fb1569007283e1e093dd8f30b7e801629/overlord/snapstate/snapstate.go#L1601) and from what I understand, `names` and `revOpts` are being linked with indices.

The problem is that [here](https://github.com/snapcore/snapd/blob/0a39bc4fb1569007283e1e093dd8f30b7e801629/overlord/snapstate/storehelpers.go#L634) names is being sorted before the mapping from name to `revOpts` is done, which could actually mess up the mapping.

This PR fixes the bug and a test is added that tries all permutations of a few snaps to make sure no regression happens due to bad ordering.

### CODE REVIEW

File: overlord/snapstate/snapstate_update_test.go [hunk 1]
```diff
@@ -10453,3 +10453,107 @@ func (s *snapmgrTestSuite) TestUpdateSetsRestartBoundaries(c *C) {
...
+	}
+	getRevOpts := func(names []string) (revOpts []*snapstate.RevisionOptions) {
+		for _, name := range names {
+			revOpts = append(revOpts, nameToRevOpts[name])
+		}
+		return revOpts
+	}
+
+	testOrder := func(names []string) {
+		revOpts := getRevOpts(names)
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:10540]: shouldn't we clear requestSnapToAction here to be sure that we are seeing the modification we expect?

**Review by pedronis — APPROVED** [review]
thank you, small remark about the test

File: overlord/snapstate/snapstate_update_test.go (hunk 1 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapstate_update_test.go:10540]: Sure, it won't hurt to make sure but I think that [the creation of a new slice](https://github.com/snapcore/snapd/pull/13265/files#diff-db72bd4eacd5472981461fb5bf207faae189669cde682db440c929328a9d13abR10505) per request is already doing this
```go
requestSnapToAction = make(map[string]*store.SnapAction, len(sa))
```

Comment (ZeyadYasser) [reply overlord/snapstate/snapstate_update_test.go:10541]: updated, thank you!

Comment (pedronis) [reply overlord/snapstate/snapstate_update_test.go:10541]: sorry, I missed the relevant code

**Review by MiguelPires — APPROVED** [review]
LGTM, thanks

File: overlord/snapstate/snapstate_update_test.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/snapstate_update_test.go:10541]: but is still better this way, as in theory the request might not happen

Comment (ZeyadYasser) [reply overlord/snapstate/snapstate_update_test.go:10541]: yes, this is definitely possible, great catch, thank you!

**Pull request merged**